File: tests/wire/wire.test.ts
// The TypeScript codec against the shared golden vectors (the Go codec runs
// the same file in server/internal/wire/wire_test.go), plus a randomized
// canonical-encoding check.
import { readFileSync } from "node:fs";
import path from "node:path";
import { test } from "node:test";
import assert from "node:assert/strict";
import { decode, encode, maxPayload, newRpcId, MalformedPacketError, type Packet } from "../../client/src/wire.ts";

interface Vectors {
  valid: { name: string; hex: string; packet: Record<string, string | number> }[];
  invalid: { name: string; hex: string }[];
}

const vectors: Vectors = JSON.parse(
  readFileSync(path.resolve(import.meta.dirname, "../../testdata/wire/vectors.json"), "utf8"),
);

const fromHex = (h: string) => Uint8Array.from(Buffer.from(h, "hex"));
const toHex = (b: Uint8Array) => Buffer.from(b).toString("hex");

function packetFromJSON(m: Record<string, string | number>): Packet {
  const rpcId = BigInt("0x" + m.rpcId);
  switch (m.type) {
    case "REQ":
      return { type: "REQ", rpcId, initialGrant: m.initialGrant as number, assetId: m.assetId as string };
    case "DATA":
      return { type: "DATA", rpcId, totalSize: m.totalSize as number, offset: m.offset as number, payload: fromHex(m.payload as string) };
    case "GRANT":
      return { type: "GRANT", rpcId, maxOffset: m.maxOffset as number, priority: m.priority as number };
    case "RESEND":
      return { type: "RESEND", rpcId, start: m.start as number, end: m.end as number };
    case "ERROR":
      return { type: "ERROR", rpcId, code: m.code as number };
  }
  throw new Error(`unknown vector type ${m.type}`);
}

for (const v of vectors.valid) {
  test(`valid: ${v.name}`, () => {
    const want = packetFromJSON(v.packet);
    assert.deepEqual(decode(fromHex(v.hex)), want);
    assert.equal(toHex(encode(want)), v.hex);
  });
}

for (const v of vectors.invalid) {
  test(`invalid: ${v.name}`, () => {
    assert.throws(() => decode(fromHex(v.hex)), MalformedPacketError);
  });
}

test("decode reads a packet at a non-zero byteOffset", () => {
  const v = vectors.valid.find((x) => x.name === "grant_basic")!;
  const inner = fromHex(v.hex);
  const outer = new Uint8Array(inner.length + 3);
  outer.set(inner, 3);
  assert.deepEqual(decode(outer.subarray(3)), packetFromJSON(v.packet));
});

test("encode rejects what decode rejects", () => {
  const rpcId = 1n;
  const bad: Packet[] = [
    { type: "REQ", rpcId, initialGrant: 0, assetId: "" },
    { type: "REQ", rpcId, initialGrant: 0, assetId: "x".repeat(0x10000) },
    { type: "REQ", rpcId, initialGrant: 0, assetId: "\ud800" }, // lone surrogate
    { type: "REQ", rpcId, initialGrant: 2 ** 32, assetId: "a" },
    { type: "DATA", rpcId, totalSize: 2, offset: 1, payload: new Uint8Array(2) },
    { type: "GRANT", rpcId, maxOffset: -1, priority: 0 },
    { type: "GRANT", rpcId, maxOffset: 1.5, priority: 0 },
    { type: "GRANT", rpcId, maxOffset: 0, priority: 256 },
    { type: "GRANT", rpcId: 2n ** 64n, maxOffset: 0, priority: 0 },
    { type: "RESEND", rpcId, start: 5, end: 5 },
  ];
  for (const p of bad) assert.throws(() => encode(p), MalformedPacketError, JSON.stringify(p, (_, x) => (typeof x === "bigint" ? `${x}n` : x)));
});

test("maxPayload leaves room for the 17-byte DATA header", () => {
  assert.equal(maxPayload(1024), 1007);
  assert.equal(maxPayload(10), 0);
});

test("newRpcId returns distinct u64 values", () => {
  const ids = new Set(Array.from({ length: 1000 }, newRpcId));
  assert.equal(ids.size, 1000);
  for (const id of ids) assert.ok(id >= 0n && id < 2n ** 64n);
});

// Same property as Go's FuzzDecode: never an unexpected exception, and anything
// accepted re-encodes to the identical bytes. Inputs are valid vectors with
// random byte mutations plus fully random buffers, from a fixed seed.
test("randomized: decode is total and canonical", () => {
  let seed = 0x1234_5678;
  const rand = () => ((seed = (Math.imul(seed, 1103515245) + 12345) >>> 0) / 2 ** 32);
  const seeds = vectors.valid.map((v) => fromHex(v.hex));
  let accepted = 0;
  for (let i = 0; i < 200_000; i++) {
    let b: Uint8Array;
    if (i % 2 === 0) {
      b = Uint8Array.from(seeds[Math.floor(rand() * seeds.length)]!);
      const flips = 1 + Math.floor(rand() * 3);
      for (let f = 0; f < flips && b.length > 0; f++) b[Math.floor(rand() * b.length)] = Math.floor(rand() * 256);
    } else {
      b = Uint8Array.from({ length: Math.floor(rand() * 40) }, () => Math.floor(rand() * 256));
    }
    let p: Packet;
    try {
      p = decode(b);
    } catch (e) {
      assert.ok(e instanceof MalformedPacketError, `unexpected ${e} for ${toHex(b)}`);
      continue;
    }
    accepted++;
    assert.equal(toHex(encode(p)), toHex(b));
  }
  assert.ok(accepted > 1000, `only ${accepted} inputs decoded; generator is too weak to mean anything`);
});

File: server/internal/wire/wire_test.go
package wire

import (
	"bytes"
	"encoding/hex"
	"encoding/json"
	"errors"
	"os"
	"reflect"
	"strconv"
	"testing"
)

// vectors mirrors testdata/wire/vectors.json, shared with the TypeScript codec.
type vectors struct {
	Valid []struct {
		Name   string         `json:"name"`
		Hex    string         `json:"hex"`
		Packet map[string]any `json:"packet"`
	} `json:"valid"`
	Invalid []struct {
		Name string `json:"name"`
		Hex  string `json:"hex"`
	} `json:"invalid"`
}

func loadVectors(t testing.TB) vectors {
	t.Helper()
	raw, err := os.ReadFile("../../../testdata/wire/vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var v vectors
	if err := json.Unmarshal(raw, &v); err != nil {
		t.Fatal(err)
	}
	return v
}

func mustHex(t testing.TB, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatal(err)
	}
	return b
}

// packetFromJSON builds the expected packet from a vector's JSON description.
func packetFromJSON(t *testing.T, m map[string]any) Packet {
	t.Helper()
	id, err := strconv.ParseUint(m["rpcId"].(string), 16, 64)
	if err != nil {
		t.Fatal(err)
	}
	rpc := RPCID(id)
	u32 := func(k string) uint32 { return uint32(m[k].(float64)) }
	switch m["type"] {
	case "REQ":
		return &Req{RPCID: rpc, InitialGrant: u32("initialGrant"), AssetID: m["assetId"].(string)}
	case "DATA":
		return &Data{RPCID: rpc, TotalSize: u32("totalSize"), Offset: u32("offset"), Payload: mustHex(t, m["payload"].(string))}
	case "GRANT":
		return &Grant{RPCID: rpc, MaxOffset: u32("maxOffset"), Priority: uint8(u32("priority"))}
	case "RESEND":
		return &Resend{RPCID: rpc, Start: u32("start"), End: u32("end")}
	case "ERROR":
		return &Error{RPCID: rpc, Code: ErrorCode(u32("code"))}
	case "META":
		fields := []Field{}
		for _, f := range m["fields"].([]any) {
			kv := f.([]any)
			fields = append(fields, Field{Name: kv[0].(string), Value: kv[1].(string)})
		}
		return &Meta{RPCID: rpc, Fields: fields}
	case "HELLO":
		return &Hello{RPCID: rpc, Caps: u32("caps")}
	case "DATA_SEQ":
		d := Data{RPCID: rpc, TotalSize: u32("totalSize"), Offset: u32("offset"), Payload: mustHex(t, m["payload"].(string))}
		return &DataSeq{Data: d, Seq: u32("seq")}
	case "RESEND_SEQ":
		return &ResendSeq{RPCID: rpc, Start: u32("start"), End: u32("end")}
	}
	t.Fatalf("unknown vector type %v", m["type"])
	return nil
}

func TestVectorsValid(t *testing.T) {
	for _, v := range loadVectors(t).Valid {
		t.Run(v.Name, func(t *testing.T) {
			want := packetFromJSON(t, v.Packet)
			got, err := Decode(mustHex(t, v.Hex))
			if err != nil {
				t.Fatalf("decode: %v", err)
			}
			if !reflect.DeepEqual(got, want) {
				t.Errorf("decode = %+v, want %+v", got, want)
			}
			enc, err := Marshal(want)
			if err != nil {
				t.Fatalf("encode: %v", err)
			}
			if h := hex.EncodeToString(enc); h != v.Hex {
				t.Errorf("encode = %s, want %s", h, v.Hex)
			}
		})
	}
}

func TestVectorsInvalid(t *testing.T) {
	for _, v := range loadVectors(t).Invalid {
		t.Run(v.Name, func(t *testing.T) {
			p, err := Decode(mustHex(t, v.Hex))
			if !errors.Is(err, ErrMalformed) {
				t.Errorf("decode = %+v, %v; want ErrMalformed", p, err)
			}
		})
	}
}

func TestEncodeRejectsWhatDecodeRejects(t *testing.T) {
	for name, p := range map[string]Packet{
		"empty asset id":      &Req{AssetID: ""},
		"asset id > 65535":    &Req{AssetID: string(make([]byte, 0x10000))},
		"non-UTF-8 id":        &Req{AssetID: "\xff"},
		"data past total":     &Data{TotalSize: 2, Offset: 1, Payload: []byte{1, 2}},
		"empty resend":        &Resend{Start: 5, End: 5},
		"meta bad name":       &Meta{Fields: []Field{{"server", "go"}}},
		"meta duplicate":      &Meta{Fields: []Field{{"etag", `"a"`}, {"etag", `"b"`}}},
		"meta empty value":    &Meta{Fields: []Field{{"etag", ""}}},
		"meta non-ASCII":      &Meta{Fields: []Field{{"content-type", "é"}}},
		"data_seq past total": &DataSeq{Data: Data{TotalSize: 2, Offset: 1, Payload: []byte{1, 2}}, Seq: 7},
		"empty resend_seq":    &ResendSeq{Start: 9, End: 9},
	} {
		if _, err := Marshal(p); !errors.Is(err, ErrMalformed) {
			t.Errorf("%s: err = %v, want ErrMalformed", name, err)
		}
	}
}

func TestMaxPayload(t *testing.T) {
	if got := MaxPayload(1024); got != 1007 {
		t.Errorf("MaxPayload(1024) = %d, want 1007", got)
	}
	if got := MaxPayload(10); got != 0 {
		t.Errorf("MaxPayload(10) = %d, want 0", got)
	}
}

// FuzzDecode: Decode never panics, and anything it accepts re-encodes to the
// identical bytes, so there is exactly one encoding per packet.
func FuzzDecode(f *testing.F) {
	v := loadVectors(f)
	for _, x := range v.Valid {
		f.Add(mustHex(f, x.Hex))
	}
	for _, x := range v.Invalid {
		f.Add(mustHex(f, x.Hex))
	}
	f.Fuzz(func(t *testing.T, b []byte) {
		p, err := Decode(b)
		if err != nil {
			return
		}
		enc, err := Marshal(p)
		if err != nil {
			t.Fatalf("decoded %+v but cannot re-encode: %v", p, err)
		}
		if !bytes.Equal(enc, b) {
			t.Fatalf("re-encode differs:\n in  %x\n out %x", b, enc)
		}
	})
}
